lms/views/api/canvas_studio.py

```diff
@@ -0,0 +1,118 @@
+"""
+Views for authorizing with Canvas Studio and listing videos.
+
+See `CanvasStudioService` for more details.
+"""
+
+from pyramid.httpexceptions import HTTPFound
+from pyramid.view import view_config
+
+from lms.security import Permissions
+from lms.services import CanvasStudioService
+from lms.validation.authentication import OAuthCallbackSchema
+
+
+# View for authorization popup which redirects to Canvas Studio's OAuth
+# authorization endpoint.
+#
+# Canvas Studio's OAuth implementation has several issues which complicate
+# using the standard OAuth flow:
+#
+# 1. Canvas Studio does not return the `state` parameter to us when it redirects
+#    after a successful authorization [1].
+# 2. Canvas Studio does not allow the use of unencrypted HTTP or `localhost`
+#    as a redirect URL for an OAuth client.
+#
+# To work around these issues, we use the following authentication flow:
+#
+# 1. User clicks button in UI to initiate Canvas Studio authorization
+# 2. The button opens `/api/canvas_studio/oauth/authorize` in a popup
+# 3. In development only, the popup redirects to change the host from
+#    http://localhost:8001 to https://hypothesis.local:48001/, keeping the
+#    path the same.
+# 4. The popup returns a 200 response which sets a cookie and does a client-side
+#    redirect to the Canvas Studio authorization endpoint.
+#
+#    We use a client-side redirect because setting cookies as part of a 302
+#    redirect response does not work in some browsers (including Chrome).
+# 5. The user completes authorization in the Canvas Studio UI
+# 6. Canvas Studio redirects to our `/api/canvas_studio/oauth/callback`
+#    endpoint, passing the authorization code, but not the `state`.
+# 7. We read the `state` parameter out of the cookie and then
+#    redirect to the `/api/canvas_studio/oauth/callback` URL that Canvas Studio
+#    should have used.
```

Comment:
The idea here is that when Canvas Studio adds the `state` parameter to the callback, we can just remove the `request.response.set_cookie` call and `oauth2_redirect_missing_state` views below.